From stable-bounces@linux.kernel.org Wed Jul 23 02:21:43 2008
From: Mohammed Gamal <m.gamal005@gmail.com>
Date: Wed, 23 Jul 2008 11:43:41 +0300
Subject: KVM: x86 emulator: Fix HLT instruction
To: stable@kernel.org
Cc: Mohammed Gamal <m.gamal005@gmail.com>
Message-ID: <1216802623-31090-6-git-send-email-avi@qumranet.com>


From: Mohammed Gamal <m.gamal005@gmail.com>

Original-Commit-Hash: bcc542267538e9ba933d08b4cd4ebd796e03a3d7

This patch fixes issue encountered with HLT instruction
under FreeDOS's HIMEM XMS Driver.

The HLT instruction jumped directly to the done label and
skips updating the EIP value, therefore causing the guest
to spin endlessly on the same instruction.

The patch changes the instruction so that it writes back
the updated EIP value.

Signed-off-by: Mohammed Gamal <m.gamal005@gmail.com>
Signed-off-by: Avi Kivity <avi@qumranet.com>
Signed-off-by: Greg Kroah-Hartman <gregkh@suse.de>

---
 arch/x86/kvm/x86_emulate.c |    2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

--- a/arch/x86/kvm/x86_emulate.c
+++ b/arch/x86/kvm/x86_emulate.c
@@ -1666,7 +1666,7 @@ special_insn:
 		break;
 	case 0xf4:              /* hlt */
 		ctxt->vcpu->arch.halt_request = 1;
-		goto done;
+		break;
 	case 0xf5:	/* cmc */
 		/* complement carry flag from eflags reg */
 		ctxt->eflags ^= EFLG_CF;
